### PR TITLE
Speed up LexicalSyntacticFeaturizer when using window features

### PR DESCRIPTION
**Pull request by alopez**
**Proposed changes**:
- Memoizes feature computations when window features are used are used in LexicalSyntacticFeaturizer. This results in some minor speedups since features are no longer computed repeatedly for each token., as noted by https://github.com/RasaHQ/rasa/issues/6406. It doesn't completely solve their problem, since more features still require more time to compute (e.g. since there are larger arrays in _features_to_onehot).

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: tests/nlu/featurizers/test_lexical_syntactic_featurizer.py [hunk 1]
```diff
@@ -195,3 +195,93 @@ def test_text_featurizer_using_pos_with_action_text(
...
+                    "pos",
+                    "upper",
+                    "digit",
+                ],
+                ["prefix2", "suffix1", "suffix3", "pos", "upper"],
+            ],
+        )
+    ],
+)
+def test_text_featurizer_using_feature_to_idx_dict(
```
Comment (dakshvar22) [inline tests/nlu/featurizers/test_lexical_syntactic_featurizer.py:222]: Can we break this test into multiple tests since it's testing too many things at once IMO?
If we want to test the memoization implementation I suggest we write a unit test just for that. To do that, it would be better if we extract out the memoization code lines 192-212 in `_tokens_to_features` to a separate function so that we can test it separately.

File: tests/nlu/featurizers/test_lexical_syntactic_featurizer.py [hunk 2]
```diff
@@ -195,3 +195,93 @@ def test_text_featurizer_using_pos_with_action_text(
...
+                        featurizer.function_dict[feature_function](tokens[position])
+                    )
+                    feature_idx = featurizer.feature_to_idx_dict[feature_name][
+                        feature_value_str
+                    ]
+                    assert token_features[feature_idx] == 1.0
+                    token_features[feature_idx] = 0.0
+
+        # Since all expected features were zeroed out, result should be all zeros
+        assert not np.any(token_features)
```
Comment (dakshvar22) [inline tests/nlu/featurizers/test_lexical_syntactic_featurizer.py:287]: I didn't fully understand what this test is doing but I think this should ideally be a separate test. Also, there is a lot of "logic" inside the test on what the code should be doing and it's a bit hard to follow. Are we testing a particular edge case here?

**Review by dakshvar22 — CHANGES_REQUESTED** [review]
Implementation is really clean! A few suggestions on how we could restructure it a bit further.

File: tests/nlu/featurizers/test_lexical_syntactic_featurizer.py (hunk 1 above)
Comment (alopez) [reply tests/nlu/featurizers/test_lexical_syntactic_featurizer.py:222]: @dakshvar22 There aren't any direct tests of the memoization: if memoization is implemented correctly, then the behavior of the feature is unchanged, but it's faster. The existing tests check behavior; I verified the speedup with cProfile. 

So technically the test is unrelated to the change. However, it exercises more parts of the code than the other tests and doesn't rely on any expected outputs, as in the other tests. By specifying those outputs, the other tests create undocumented (afaict) requirements on the order of the features, overconstraining the implementation. This test does not have any expected output. Rather, it checks whether the features are internally consistent, irrespective of how the features are ordered.

Comment (alopez) [reply tests/nlu/featurizers/test_lexical_syntactic_featurizer.py:222]: I guess then the question is whether this test should be separated out or removed altogether?

File: tests/nlu/featurizers/test_lexical_syntactic_featurizer.py (hunk 2 above)
Comment (alopez) [reply tests/nlu/featurizers/test_lexical_syntactic_featurizer.py:287]: Just so I understand: do you mean that this should be separated out from the rest of the test?

In a nutshell, the this test is checking that the features returned for the test input are 1 if and only if they were features requested in `featurizer_config`. Rather than use a hardcoded expected features (see comment above), it does this by indexing the features by name via `feature_to_idx_dict`. This is arguably a whitebox test, but afaict `feature_to_idx_dict` is public so it seems ok to test it.

File: rasa/nlu/featurizers/sparse_featurizer/lexical_syntactic_featurizer.py [hunk 3]
```diff
@@ -189,14 +189,34 @@ def _tokens_to_features(self, tokens: List[Token]) -> List[Dict[Text, Any]]:
...
+            ]
+        )
+        token_to_function_values = []
+        for token_idx, token in enumerate(tokens):
+            function_values = {}
+            for function in functions:
+                function_values[function] = self._get_feature_value(
+                    function, token, token_idx, len(tokens)
+                )
+            token_to_function_values.append(function_values)
```
Comment (dakshvar22) [inline rasa/nlu/featurizers/sparse_featurizer/lexical_syntactic_featurizer.py:210]: I think it would also help to move this memoization bit to an independent function. Would be easier to follow IMO

File: tests/nlu/featurizers/test_lexical_syntactic_featurizer.py (hunk 2 above)
Comment (dakshvar22) [reply tests/nlu/featurizers/test_lexical_syntactic_featurizer.py:287]: As I said above, having this as a complete test is still useful, so let's keep it.
